src/native/s3select/s3select_napi.cpp

```diff
@@ -0,0 +1,277 @@
+/* Copyright (C) 2016 NooBaa */
+
+#include "../../../src/native/third_party/isa-l/include/crc.h"
+#include "../../../submodules/s3select/include/s3select.h"
+#include "../util/common.h"
+#include "../util/napi.h"
+#include <arpa/inet.h>
+
+namespace noobaa
+{
+
+const char *CSV_FORMAT = "CSV";
+const char *JSON_FORMAT = "JSON";
+const char *PARQUET_FORMAT = "PARQUET";
+
+class S3SelectNapi;
+
+class SelectWorker : public Napi::AsyncWorker
+{
+public:
+    Napi::ObjectReference _args_ref;
+    Napi::Promise::Deferred _deferred;
+    Napi::ObjectWrap<S3SelectNapi>& _wrap;
+    s3selectEngine::csv_object *_csv_object;
+    s3selectEngine::json_object *_json_object;
+    std::string _select;
+    const char* _buffer;
+    const unsigned int _buffer_len;
+    const bool _is_flush;
+    const std::string _input_format;
+    const uint8_t* _headers_bytes;
+    const uint32_t _headers_len;
+
+    SelectWorker(const Napi::CallbackInfo& info,
+        Napi::ObjectWrap<S3SelectNapi>& wrap,
+        s3selectEngine::csv_object* csv_object,
+        s3selectEngine::json_object* json_object,
+        const char* buffer,
+        const unsigned int buffer_len,
+        const bool is_flush,
+        const std::string input_format,
+        const uint8_t* headers_bytes,
+        const uint32_t headers_len)
+        : AsyncWorker(info.Env())
+        , _args_ref(Napi::Persistent(Napi::Object::New(info.Env())))
+        , _wrap(wrap)
+        , _csv_object(csv_object)
+        , _json_object(json_object)
+        , _buffer(buffer)
+        , _buffer_len(buffer_len)
+        , _is_flush(is_flush)
+        , _input_format(input_format)
+        , _deferred(info.Env())
+        , _headers_bytes(headers_bytes)
+        , _headers_len(headers_len)
+    {
+        //take a ref on args to make sure they are not GCed until worker is done
+        uint32_t i;
+        for (i = 0; i < info.Length(); ++i) {
+            _args_ref.Set(i, info[i]);
+        }
+        _args_ref.Set(i, _headers_bytes);
```

Comment:
This won't work, you have to keep the ref to the original js buffer, not to the memory pointer...

Reply:
Moved to S3Select_Napi ctor.